### PR TITLE
ENH: Hdf5 update

### PR DESCRIPTION
**Pull request by hjmjohnson**
Update to most recent tagged version of HDF5 (hdf5-1_10_6) that is compatible with cmake 3.10.2.

Replace customizations for ITK that are now included in the upstream version of HDF5.

## PR Checklist
- [X] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [X] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)

### CODE REVIEW

**Review by dzenanz — APPROVED** [review]

**Comment by hjmjohnson** [comment]
@dzenanz There is a windows failure.  It looks like a small problem for someone who has ready access to windows.

https://open.cdash.org/viewBuildError.php?buildid=6920225

Could you help identify what needs to be fixed?

**Comment by dzenanz** [comment]
I don't run into those compile errors with either DLLs or static libraries. Both my builds have ITK_LEGACY_REMOVE on.

**Comment by dzenanz** [comment]
But based on the error message, it is probably this: https://stackoverflow.com/questions/3491990/c-definition-of-dllimport-static-data-member

**Comment by dzenanz** [comment]
I now tried with `BUILD_SHARED_LIBS` on, and both variants of `ONLY_SHARED_LIBS`. I cannot reproduce the compile error.

**Comment by hjmjohnson** [comment]
@dzenanz oh-boy,  If you can not figure out how to reproduce the error, I'm going to be really out of luck :)

Any suggestions on what to try next?

**Comment by dzenanz** [comment]
Bisect different HDF5 versions? But before that, have you seen [Azure warnings](https://open.cdash.org/build/6920843/configure)?
```text
-- Checking IF alignment restrictions are strictly enforced... yes
CMake Warning (dev) at Modules/ThirdParty/HDF5/src/itkhdf5/CMakeLists.txt:438 (option):
  Policy CMP0077 is not set: option() honors normal variables.  Run "cmake
  --help-policy CMP0077" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  For compatibility with older versions of CMake, option is clearing the
  normal variable 'ONLY_SHARED_LIBS'.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Warnings Configuration:
-- Filter ZLIB is ON
```
GitHub actions also use Azure, and delegate to `itk_common.cmake`. But I don't see any special options in the [initial cache](https://github.com/InsightSoftwareConsortium/ITK/blob/dashboard/itk_common.cmake#L366-L376). I will also try a clean build.

**Comment by dzenanz** [comment]
A clean build with static libraries was successful. Trying shared.

**Comment by dzenanz** [comment]
Reconfiguring to `BUILD_SHARED_LIBS` also built successfully.

**Comment by dzenanz** [comment]
@N-Dekker @jhlegarreta can you try reproducing/fixing the error?

**Comment by jhlegarreta** [comment]
Which set of flags/CMake config options should we try? By default, I have `ITK_LEGACY_REMOVE` to `OFF` and I build static libraries (`BUILD_SHARED_LIBS` to `OFF`) on a Windows 10 machine, MSVC 2015.

**Comment by jhlegarreta** [comment]
> Which set of flags/CMake config options should we try? By default, I have ITK_LEGACY_REMOVE to OFF and I build static libraries (BUILD_SHARED_LIBS to OFF) on a Windows 10 machine, MSVC 2015.

Result with the above settings: 
`Build: 331 succeeded, 2 failed, 93 up-to-date, 9 skipped`
Failure for `itkPasteImageFilterGTest.cxx`

So unrelated to this patch set.

**Comment by hjmjohnson** [comment]
> > Which set of flags/CMake config options should we try? By default, I have ITK_LEGACY_REMOVE to OFF and I build static libraries (BUILD_SHARED_LIBS to OFF) on a Windows 10 machine, MSVC 2015.
> 
> Result with the above settings and commit hash [6813461](https://github.com/InsightSoftwareConsortium/ITK/commit/681346119f8bc823857065206c1ad42b73aa4c8f):
> `Build: 331 succeeded, 2 failed, 93 up-to-date, 9 skipped`
> Failure for `itkPasteImageFilterGTest.cxx`
> 
> So unrelated to this patch set.
> 
> Sorry @hjmjohnson , but I had started the build before your latest patches. Let me know if I should try again with a given commit.

I had accidentally committed unrelated file changes.  This has been fixed.  I also fixed cmake warning. 

The failing dashboard entry is https://open.cdash.org/build/6922596/notes#note0

**Comment by jhlegarreta** [comment]
@hjmjohnson Do you still need me to try to reproduce the issue/see if the build is successful on my machine on the latest patch set?

**Comment by dzenanz** [comment]
@jhlegarreta I understood yes. The new commit just changes the error message?

**Comment by jhlegarreta** [comment]
> @jhlegarreta I understood yes. The new commit just changes the error message?

I clean/rebuilt all without modifying my settings (let me know if I should use the ones in the `CDash` build) and the build failed:
```
Rebuild All: 415 succeeded, 10 failed, 9 skipped
```
The error report in VS reads
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK1104	cannot open file '..\..\..\..\lib\Debug\itkhdf5-static-5.2_D.lib'	itkAlgorithmsPrintTest2	C:\SDKs\ITK\itk-head-Bin\Modules\Nonunit\IntegratedTest\test\LINK	1	
Error	C1083	Cannot open source file: 'C:\SDKs\ITK\itk-head-Bin\Modules\ThirdParty\HDF5\src\itkhdf5\H5Tinit.c': No such file or directory	hdf5-static	C:\SDKs\ITK\itk-head-Bin\Modules\ThirdParty\HDF5\src\itkhdf5\src\c1	1	
Error	LNK1104	cannot open file '..\..\..\..\lib\Debug\itkhdf5-static-5.2_D.lib'	itkAlgorithmsPrintTest	C:\SDKs\ITK\itk-head-Bin\Modules\Nonunit\IntegratedTest\test\LINK	1	
Error	LNK1104	cannot open file '..\..\..\..\lib\Debug\itkhdf5-static-5.2_D.lib'	itkAlgorithmsPrintTest4	C:\SDKs\ITK\itk-head-Bin\Modules\Nonunit\IntegratedTest\test\LINK	1	
Error	LNK1104	cannot open file '..\..\..\..\lib\Debug\itkhdf5-static-5.2_D.lib'	itkAlgorithmsPrintTest3	C:\SDKs\ITK\itk-head-Bin\Modules\Nonunit\IntegratedTest\test\LINK	1	
Error	LNK1104	cannot open file '..\..\..\..\lib\Debug\itkhdf5-static-5.2_D.lib'	ITKIOMINCTestDriver	C:\SDKs\ITK\itk-head-Bin\Modules\IO\MINC\test\LINK	1	
Error	LNK1104	cannot open file '..\..\..\..\lib\Debug\itkhdf5-static-5.2_D.lib'	ITKIOHDF5TestDriver	C:\SDKs\ITK\itk-head-Bin\Modules\IO\HDF5\test\LINK	1	
Error	LNK1104	cannot open file '..\..\..\..\lib\Debug\itkhdf5-static-5.2_D.lib'	ITKIOTransformHDF5TestDriver	C:\SDKs\ITK\itk-head-Bin\Modules\IO\TransformHDF5\test\LINK	1	
Error	LNK1104	cannot open file '..\..\..\..\lib\Debug\itkhdf5-static-5.2_D.lib'	ITKIntegratedTestTestDriver	C:\SDKs\ITK\itk-head-Bin\Modules\Nonunit\IntegratedTest\test\LINK	1	
```

(I am additionally getting a failure in `itkPasteImageFilterGTest.cxx` as yesterday but that is not relevant for this patch set/had observed it before).

Are further details needed? Let me know if I should configure an experimental build so that the next builds are submitted to the dashboard.

**Comment by hjmjohnson** [comment]
I think I figured out the problem.  WOrking on a solution (on my mac) so hopefully windows tests will start passing.

**Comment by hjmjohnson** [comment]
@jhlegarreta If you could try building a shared library version of this patch on windows.  

I'm pretty sure the old issue occurred because ITK was being built with shared libraries, and HDF5 was trying to build with both shared and static library versions.  ITK is not instrumented to do that.  I needed to add  `set(ONLY_SHARED_LIBS ON CACHE INTERNAL "Only build shared")` in the correct location for it to be recognized and respected.  

At least that is what I hope was the problem.

**Comment by jhlegarreta** [comment]
@hjmjohnson thanks for investigating this. Is commit 3989fc8 the one to try with a shared build?

**Review by thewtex — APPROVED** [review]

File: Modules/ThirdParty/HDF5/src/itkhdf5/CMakeLists.txt [hunk 1]
```diff
@@ -7,6 +7,10 @@ endif()
 # ITK --stop
 
 
+if(POLICY CMP0074)
```
Comment (dzenanz) [inline Modules/ThirdParty/HDF5/src/itkhdf5/CMakeLists.txt:10]: The same thing is done just below on line 14.

Comment (hjmjohnson) [reply Modules/ThirdParty/HDF5/src/itkhdf5/CMakeLists.txt:9]: Fixed.

**Comment by hjmjohnson** [comment]
@dzenanz @thewtex  I don't know how to make the ghost-flow test to pass.  I tried everything I could think of.

**Comment by dzenanz** [comment]
ghostflow-check-master still fails. Did you rebase on master after I merged the limit increase?

**Comment by hjmjohnson** [comment]
@dzenanz Yes.

**Comment by dzenanz** [comment]
@bradking might be able to explain the problem and/or fix it?

**Comment by bradking** [comment]
@hjmjohnson @dzenanz the limit needs to be set via a `hooks-max-size` attribute set on relevant sources.  I added a commit to set it on the specific sources in question.

**Comment by jhlegarreta** [comment]
Shared build corresponding to commit 0e8f77c was successful on my platform/machine (Windows 10, MSVC 2015).

File: Modules/ThirdParty/HDF5/.gitattributes [hunk 2]
```diff
@@ -0,0 +1,4 @@
+src/itkhdf5/src/H5C.c        hooks-max-size=500000
```
Comment (hjmjohnson) [inline Modules/ThirdParty/HDF5/.gitattributes:1]: OH!  I added this in ~/ITK/.gitattributes

**Comment by hjmjohnson** [comment]
> Shared build corresponding to commit [0e8f77c](https://github.com/InsightSoftwareConsortium/ITK/commit/0e8f77c47ab7f65404adfa9616bd24f6cd0c8349) was successful on my platform/machine (Windows 10, MSVC 2015).

@jhlegarreta  WOOT WOOT!  We are getting very close!  Now we just need to wait several hours for the builds to replicate the successes.  Thank you for your efforts.

**Comment by seanm** [comment]
Hi guys. Some change, I'm guessing this PR, seems to have changed the installed HDF library files names from:

- libitkhdf5_debug.a 
- libitkhdf5_cpp_debug.a

to:

- libitkhdf5-static_debug.a
- libitkhdf5_cpp-static_debug.a

I have nightly builds of my own app against nightly builds of ITK, and this renaming broke it today.

Is the rename deliberate?  Could it be reverted?  Thanks.

**Comment by thewtex** [comment]
Hi @seanm,

`-static` was deliberately added in https://github.com/InsightSoftwareConsortium/ITK/pull/2229 to follow upstream and address https://github.com/InsightSoftwareConsortium/ITK/issues/1971. Why is this breaking your build?

**Comment by seanm** [comment]
It breaks my build the same way renaming an API would. My Xcode project refers to the libraries it links against by their filename, so if the filename changes, the library is not found, and linking fails.  I have nightly builds against a stable ITK release but also against nightly ITK.  My workaround is to rename them back when building nightly ITK.

I guess since it's deliberate I'll live with it.  On its face though, it seems unnecessary to add the word "static" when the extension already indicates it's a static lib, and no other itk libs have the word "static" in them.

**Pull request merged**